sdk/storage/azure-storage-common/src/xml_wrapper.cpp

```diff
@@ -127,18 +127,25 @@ namespace Azure { namespace Storage { namespace Details {
     xmlBufferFree(static_cast<xmlBufferPtr>(m_buffer));
   }
 
+  namespace {
+    template <typename T> inline constexpr xmlChar* BadCast(T const* x)
+    {
+      return const_cast<xmlChar*>(reinterpret_cast<xmlChar const*>(x));
+    }
+  } // namespace
+
```

Comment:
Did you do this to avoid some kind of compiler warning?

Reply:
Yes, to avoid c-style cast. The macro that was used, `BAD_CAST` from libxml2, extends to a c-style cast, and I enabled a warning for the compiler to warn when it is used, so we use one of the proper C++-style casts.

Reply:
Do we need template here? Isn't it always const char*?

Reply:
You don't have to `inline` a template function. But if you mean it, it's okay.

Reply:
We use `const Type*` rather than `Type const*`, be consistent

Reply:
I rather meant to inline, to come to macro as close as possible. It will be a no-op in the generated code anyways, just a cast to satisfy the compiler. LMK if you'd like to see it non-inlined. But I'd recommend to explicitly specify it as inline.

Reply:
Interestingly, if you make it non-template, GCC complains it can't be a constexpr. Must be a bug, but still, I'd rather have guaranteed constexpr.

Reply:
Why are you so obsessed with `constexpr` here? `reinrepret_cast` and `const_cast` are almost always no-op when translated to assembly code. So I think it doesn't matter to be `constexpr` or not.

Reply:
https://stackoverflow.com/questions/26200827/getting-around-the-reinterpret-cast-limitation-with-constexpr

I found a discussion here. I don't know whether it's defined behavior to do `reinterpret_cast` in `constexpr` function. But I would avoid doing that if I'm not absolutely sure it's safe.

Reply:
My point is: we use template when we need template, we don't use template to get around some compiler error caused by combining `reinterpret_cast` and `constexpr`.

I just checked, we use all those `BAD_CAST` for `const char*`, so I don't think we need template here at all.